pkg/clusterapi/controlplane.go

```diff
@@ -0,0 +1,125 @@
+package clusterapi
+
+import (
+	"context"
+
+	etcdv1 "github.com/mrajashree/etcdadm-controller/api/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+
+	"github.com/aws/eks-anywhere/pkg/clients/kubernetes"
+)
+
+// ProviderControlPlane represents the provider specific part of a CAPI control plane definition
+type ProviderControlPlane interface {
+	// Objects returns all provider specific API objects
+	Objects() []kubernetes.Object
+}
+
+// NoObjectsProviderControlPlane is a helper for providers that don't require extra objects in the control plane
+type NoObjectsProviderControlPlane struct{}
+
+func (NoObjectsProviderControlPlane) Objects() []kubernetes.Object {
+	return nil
+}
+
+// Object represents a kubernetes API object
+type Object interface {
+	kubernetes.Object
+}
+
+// ControlPlane represents the spec for a CAPI control plane for an specific provider
+type ControlPlane[C, M Object, P ProviderControlPlane] struct {
+	Cluster                     *clusterv1.Cluster
+	ProviderCluster             C
+	KubeadmControlPlane         *controlplanev1.KubeadmControlPlane
+	ControlPlaneMachineTemplate M
+	EtcdMachineTemplate         M
+	EtcdCluster                 *etcdv1.EtcdadmCluster
+
+	// Provider holds the provider specific components for the control plane
+	Provider P
+}
+
+// Objects returns all API objects that form a concrete control plane for an specific provider
+func (cp *ControlPlane[C, M, P]) Objects() []kubernetes.Object {
+	objs := cp.Provider.Objects()
+	objs = append(objs, cp.Cluster, cp.KubeadmControlPlane)
+	if cp.EtcdCluster != nil {
+		objs = append(objs, cp.EtcdCluster)
+	}
+
+	return objs
+}
+
+// UpdateImmutableObjectNames checks if any control plane immutable objects have changed y comparing the new definition
+// with the current state of the cluster. If they had, it generates a new name for them by increasing a monotonic number
+// at the end of the name
+// This is applied to all provider machine templates
+// machineTemplateComparator receives the new definition of a machine template with the name of the current one
```

Comment:
this is awesome!
wonder if we want to be more generic like `capiObjectComparator` since there are cases we need new `KubeadmConfigTemplate` or potentially more objects need comparison